cscs-checks/mch/src/cuda_stencil_test.cu

```diff
@@ -0,0 +1,218 @@
+#include "cuda.h"
+#include <cassert>
+#include <chrono>
+#include <iostream>
+#include <stdio.h>
+
+#define N 800
+#define TPB 16
+#define RED_NB 50
+
+#define BigN 4000
+#define BigTPB 20
+
+#define memN 200000000
+#define memTPB 500
+
+using namespace std;
+using chrono_clock = std::chrono::high_resolution_clock;
+using sec_dur = std::chrono::duration<double, std::ratio<1, 1>>;
+
+__global__ void bigstencil(int* a) {
+  unsigned long long i = blockIdx.x * blockDim.x + threadIdx.x;
+  a[i] = 7;
+}
+
+__global__ void indexfill(int* a) {
+  int i = blockIdx.x * blockDim.x + threadIdx.x;
+  a[i] = i;
+}
+
+__global__ void dotProduct(int* a, int* b, int* c) {
+  __shared__ int temp[N];
+  int i = blockIdx.x * blockDim.x + threadIdx.x;
+  temp[i] = a[i] * b[i];
+
+  __syncthreads();
+  if(i == 0) {
+    int sum = 0;
+    for(int i = 0; i < N; i++)
+      sum += temp[i];
+
+    *c = sum;
+  }
+}
+
+__global__ void reduction(int* input, int* output) {
+  __shared__ int tmp[TPB];
+
+  tmp[threadIdx.x] = input[threadIdx.x + blockIdx.x * blockDim.x];
+
+  __syncthreads();
+
+  if(threadIdx.x < blockDim.x / 2)
+    tmp[threadIdx.x] += tmp[threadIdx.x + blockDim.x / 2];
+
+  __syncthreads();
+
+  if(threadIdx.x < blockDim.x / 4)
+    tmp[threadIdx.x] += tmp[threadIdx.x + blockDim.x / 4];
+
+  __syncthreads();
+
+  if(threadIdx.x < blockDim.x / 8)
+    tmp[threadIdx.x] += tmp[threadIdx.x + blockDim.x / 8];
+
+  __syncthreads();
+
+  if(threadIdx.x == 0) {
+    tmp[threadIdx.x] += tmp[threadIdx.x + 1];
+    output[blockIdx.x] = tmp[threadIdx.x];
+  }
+}
+__global__ void bigstencil(int* in, int* out) {
+  int i = blockIdx.x * blockDim.x + threadIdx.x;
+  out[i] = in[i] + 2;
+}
+
+int main() {
+
+  // CSCS: Initialize CUDA runtime outside measurement region
+  cudaFree(0);
```

Comment:
@lxavier I have added this line to your benchmark, because the performance measurements were not reliable, since you were also counting the CUDA runtime initialization time. The initialization overhead is much more significant on CUDA 9 than in CUDA 8.